Add filtering page (#663)
saleor/saleor-docs: docs/developer/filtering.mdx

---
title: Filtering
---

:::info

> The new filtering approach was introduced in **Saleor 3.11**. For now, it's only

Reviewer: Maybe let's call it "where filtering", what do you think?

```suggestion
The "where" filtering approach was introduced in **Saleor 3.11**. For now, it's only
```

> The "where" filtering approach was introduced in **Saleor 3.11**. For now, it's only
> available on the `Attribute` type.

Reviewer 2: ```suggestion
The `where` filtering approach was introduced in **Saleor 3.11**. For now, it's only
available on the `Attribute` type.
```

Also, regarding that:
>  For now, it's only available on the `Attribute` type.

Isn't it the [`attributes`](https://docs.saleor.io/docs/3.x/api-reference/queries/attributes) query?

Author: Like I mean here that it's done for `Attributes` type with reference to the whole idea of adding the filtering input for each `Attribute` field, and adding a new filter on `attributes` query.

:::

:::caution

This feature is currently in **Feature Preview**. This part of Saleor is not complete
and subject to change but is available to experiment and provide feedback.

:::

## Introduction

> The "where" filtering allows specifying more complex conditions with the use of `AND` and `OR` operators,

Reviewer 2: ```suggestion
The `where` filtering allows specifying more complex conditions with the use of `AND` and `OR` operators,
```

combined with operations like equal (`eq`), one of, and range. It's available under the `where` option,

> for now, only on attributes query.

Reviewer: ``` suggestion
The "where" filtering allows specifying complex conditions with the use of `AND` and `OR` operators,
combined with operations like equal (`eq`), one of, and range. It's available under the `where` option,
for now, only on the [attributes](api-reference/queries/attributes.mdx) query.
```

### General approach

- Currently only `AND` and `OR` operators are available.
  (See the example for [AND](#filtering-with-the-and-operator) and [OR](#filtering-with-the-or-operator) operator usage).

> - It's allowed to filter by flat fields like it was before, as well as specifying operators,

Reviewer: I think we shouldn't write as if this was a changelog, but as it is now.

```suggestion
- It's allowed to filter by flat fields as well as specifying operators,
```

  but this approach cannot be mixed. ([See the example](#mixing-flat-field-filter-and-operator-and-or)).
- Only one operator can be specified on each level. ([See the example](#mixing-operators-on-the-same-level)).

> - In the new approach we decided to add a filter option for each field of the given type.

Reviewer: I would rephrase this to not sound like a changelog, but a current state.

> - In the new "where" approach, each field of a given type has a corresponding filter option.

Reviewer 2: ```suggestion
- In the new `where` approach, each field of a given type has a corresponding filter option.
```

>   As a result, the "search" filter is now located at the top level. ([See the example](##search-filtering)).

Reviewer 2: ```suggestion
  As a result, the `search` filter is now located at the top level. ([See the example](##search-filtering)).
```

- The filter input type corresponds to the property name in the API type.
- The string, enum, date, and all numeric type fields have specific input types

>   with available operations options. Currently there is option to filter by specific
>   value ([example](#filtering-by-a-specific-value)) and list of values ([example](#filtering-by-a-list-of-values)).

Reviewer 2: ```suggestion
  with available operations options. Currently, there is an option to filter by specific
  value ([example](#filtering-by-a-specific-value)) and list of values ([example](#filtering-by-a-list-of-values)).
```

> - You can only specify one operation. Using eq or oneOf is acceptable, but using both will result in an error. ([See the example](#mixing-different-operations)).

Reviewer 2: ```suggestion
- You can only specify one operation. Using `eq` or `oneOf` is acceptable, but using both will result in an error. ([See the example](#mixing-different-operations)).
```

- You cannot provide both `filter` and `where` options at the same time. ([See the example](#mixing-filter-and-where-options)).

## Examples

To use new filters you need to specify the `where` option.
The following examples will be presented on the `attributes` query.

### Filtering by a specific value

To filter by a specific value, use the `eq` operation. Only object with the exact match
will be returned.

> **_Example:_** _Return attributes that have `Flavor` name._

```graphql {2}
query {
  attributes(first: 20, where: { type: { oneOf: ["Flavor"] } }) {
    edges {
      node {
        id
        name
        inputType
      }
    }
  }
}
```

### Filtering by a list of values

To filter by a list of values use the `oneOf` operator. All objects that have a matching value
specified in the list are returned.

> **_Example_**: _Return attributes that input type is `DATE`, or `DATE_TIME`._

```graphql {2}
query {
  attributes(first: 20, where: { inputType: { oneOf: [DATE, DATE_TIME] } }) {
    edges {
      node {
        id
        name
        inputType
      }
    }
  }
}
```

### Filtering with the `AND` operator

> The `AND` operator allows defining a list of conditions that must be fulfilled.

Reviewer 2: ```suggestion
The `AND` operator allows for defining a list of conditions that must be met.
```

> **_Example:_** _Return attributes that are `PRODUCT_TYPE` type, and the `id` is `QXR0cmlidXRlOjIx` or `QXR0cmlidXRlOjI3`, or the attribute `name` is `Size`._

```graphql {5-13}
query {
  attributes(
    first: 20
    where: {
      AND: [
        { type: { eq: PRODUCT_TYPE } }
        {
          OR: [
            { ids: ["QXR0cmlidXRlOjIx", "QXR0cmlidXRlOjI3"] }
            { name: { eq: "Size" } }
          ]
        }
      ]
    }
  ) {
    edges {
      node {
        id
        name
      }
    }
  }
}
```

### Search filtering

To use the full-text search capabilities, use the root-level `search` argument.

> **_Example:_** _Search attributes by `size` value_

```graphql {2}
query {
  attributes(first: 20, search: "size") {
    edges {
      node {
        id
        name
      }
    }
  }
}
```

## Invalid examples

### Mixing flat field filter and operator `AND`, `OR`

> **_Example:_** _Return attributes that name is `Author` and type is `PRODUCT-TYPE`._

```graphql {4-5}
query {
   attributes(
    where: {
        type: {eq: PRODUCT-TYPE}},
        AND: [{name: {eq: "Author"}}]
    }) {
      edges {
          node {
             id
          }
      }
   }
}
```

> **INSTEAD:** Join both conditions into one AND filter input

Reviewer 2: ```suggestion
**INSTEAD:** Join both conditions into one `AND` filter input
```
